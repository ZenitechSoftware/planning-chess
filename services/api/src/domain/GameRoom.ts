import { WebSocket } from 'ws';

export interface GameWebSocket extends WebSocket {
  roomId: string;
}
