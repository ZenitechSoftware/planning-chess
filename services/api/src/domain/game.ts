export interface Turn {
  row?: number;
  tile?: number;
  figure?: string;
  playerId: string;
  score?: number;
  turnType: TurnType;
}

export enum TurnType {
  MoveSkipped = 'MoveSkipped',
  FigurePlaced = 'FigurePlaced',
}
