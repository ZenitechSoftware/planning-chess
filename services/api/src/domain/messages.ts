import WebSocket from 'ws';

export enum MessageType {
  PlayerConnected = 'PlayerConnected',
  NewPlayer = 'NewPlayer',
  PlayerDisconnected = 'PlayerDisconnected',
  FigureMoved = 'FigureMoved',
  PlayerFinishMove = 'PlayerFinishMove',
  NewBoardState = 'NewBoardState',
  ClearBoard = 'ClearBoard',
  SetMyTurn = 'SetMyTurn',
}

export interface Message {
  type: MessageType;
  payload: unknown;
}

export interface PlaceFigureMessage {
  row: number;
  tile: number;
  figure: string;
  player: string;
}

export interface NewPlayerMessage {
  playerName: string;
}

export interface Handler {
  (ws: WebSocket, payload: unknown): void;
}
