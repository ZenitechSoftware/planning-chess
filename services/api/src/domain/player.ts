export enum PlayerStatus {
  MoveSkipped = 'MoveSkipped',
  FigurePlaced = 'FigurePlaced',
  ActionNotTaken = 'ActionNotTaken',
}

export interface Player {
  id: string;
  name: string;
  color: PlayerAvatarColor;
  status: PlayerStatus;
}

interface RGB {
  r: number;
  g: number;
  b: number;
}

export interface PlayerAvatarColor {
  background: RGB;
  text: RGB;
}
