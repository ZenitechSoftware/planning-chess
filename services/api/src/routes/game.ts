import express from 'express';
import Joi from 'joi';
import { config } from 'dotenv';

const router = express.Router();

config();

const validationOfPath = () => {
  const uuidPath = `${process.env.GAME_ROOM_UUID}`;
  return Joi.string().valid(uuidPath);
};

router.use((_req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  next();
});

router.get('/:id', async (req, res) => {
  const passedUUID = req.params.id;
  const pathValidationSchema = validationOfPath();

  await pathValidationSchema
    .validateAsync(passedUUID)
    .then(() => res.status(200).send({ message: 'Validation Successful' }))
    .catch(() => res.status(400).send({ message: 'Validation Error' }));
});

export default router;
