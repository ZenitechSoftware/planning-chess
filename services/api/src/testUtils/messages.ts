import {
  MessageType,
  PlaceFigureMessage,
  ReceivedMessage,
} from '../domain/messages';

export const getFigureMovedMessage = (
  turn: PlaceFigureMessage,
): ReceivedMessage<MessageType.FigureMoved> => ({
  type: MessageType.FigureMoved,
  payload: turn,
});

export const getMoveSkippedMessage = (
  playerId: string,
): ReceivedMessage<MessageType.MoveSkipped> => ({
  type: MessageType.MoveSkipped,
  payload: { playerId: playerId },
});
