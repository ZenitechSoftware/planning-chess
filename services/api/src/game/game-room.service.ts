import WebSocket, { WebSocketServer } from 'ws';
import { v4 as uuidv4 } from 'uuid';
import logger from '../logger';
import { Player } from '../domain';
import { PlaceFigureMessage } from '../domain/messages';

interface GameRoom {
  roomId: string;
  server: WebSocketServer;
  players: Map<WebSocket, Player>;
  turns: PlaceFigureMessage[];
}

const rooms = new Map<string, GameRoom>();

export const getOrCreateRoom = (id?: string): GameRoom => {
  const roomId = id ?? uuidv4();
  const room = rooms.get(roomId);

  if (room) {
    logger.info(`Room ${roomId} already exists.`);
    return room;
  }

  const server = new WebSocketServer({
    noServer: true,
  });

  const newRoom = {
    roomId,
    server,
    players: new Map<WebSocket, Player>(),
    turns: <PlaceFigureMessage[]>[],
  };
  rooms.set(roomId, newRoom);

  return newRoom;
};

export const getClients = (id: string): Set<WebSocket> =>
  rooms.get(id)?.server.clients;

export const getPlayers = (id: string): Map<WebSocket, Player> | null =>
  rooms.get(id)?.players || null;

export const getTurns = (id: string): Array<PlaceFigureMessage> =>
  rooms.get(id)?.turns;

export const printRooms = (): void => {
  for (const entry of rooms.entries()) {
    logger.debug(`${entry[0]} -> clients: ${entry[1].server.clients.size}`);
  }
};

export const cleanUp = (id: string): void => {
  if (getClients(id)?.size === 0) {
    logger.info(`Closing game room ${id}`);
    rooms.get(id)?.server.close();
    rooms.delete(id);
  }
};
