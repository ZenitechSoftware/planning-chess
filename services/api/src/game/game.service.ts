import { PlaceFigureMessage } from '../domain/messages';
import { calculateScore } from '../helpers/calculate-score';
import * as gameRoomService from './game-room.service';
import logger from '../logger';

export const playerHasMove = (roomId: string, playerId: string): boolean =>
  Boolean(findMoveByPlayerId(roomId, playerId));

export const playerHasSkipped = (roomId: string, playerId: string): boolean =>
  Boolean(gameRoomService.getSkippedPlayers(roomId).find(player => player.userId === playerId));

export const figureMoved = (
  roomId: string,
  payload: PlaceFigureMessage,
): PlaceFigureMessage[] => {
  let score;
  if (payload !== null) {
    score = calculateScore(payload);
  }

  if (playerHasMove(roomId, payload.id)) {
    removeTurn(roomId, payload.id);
  }

  gameRoomService.getTurns(roomId).push({ ...payload, score });
  return gameRoomService.getTurns(roomId);
};

export const clearBoard = (roomId: string): void => {
  gameRoomService.getTurns(roomId).length = 0;
};

export const findMoveByPlayerId = (
  roomId: string,
  id: string,
): PlaceFigureMessage | undefined =>
  gameRoomService.getTurns(roomId).find((turn) => turn.id === id);

export const getBoard = (roomId: string): PlaceFigureMessage[] =>
  gameRoomService.getTurns(roomId);

export const removeTurn = (roomId: string, playerId: string): void => {
  const turns = gameRoomService.getTurns(roomId);
  try {
    const turnIndex = turns.findIndex((turn) => turn.id === playerId);
    if (turnIndex < 0) {
      throw new Error(`Player with ${playerId} id move could not be found`);
    }
    turns.splice(turnIndex, 1);
  } catch (err) {
    logger.error(err?.message);
  }
};
