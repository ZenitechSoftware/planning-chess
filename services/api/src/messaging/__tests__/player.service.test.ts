import WebSocket from 'ws';
import * as playerService from '../players.service';
import { MessageType } from '../../domain/messages';
import * as gameService from '../../game/game.service';

jest.mock('ws');
jest.mock('uuid', () => ({
  v4: () => 'some-short-v4-uuid-0',
}));
jest.mock('../../game/game.service');

describe('player.service', () => {
  const ws = new WebSocket('');
  beforeAll(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(1);
    Object.defineProperty(ws, 'readyState', { value: WebSocket.OPEN });
    playerService.subscribe(ws, { playerName: 'player1' });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  afterAll(() => {
    playerService.unsubscribe(ws);
  });

  it('should join a new player', () => {
    const message = {
      type: MessageType.PlayerConnected,
      payload: { playerName: 'foo' },
    };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock.mock.calls).toMatchSnapshot();
  });

  it('should skip a move for a player', () => {
    const payload = { userId: 'some-short-v4-uuid-0' };
    const message = { type: MessageType.MoveSkipped, payload };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock.mock.calls).toMatchSnapshot();
  });

  it('should not skip a move, because user do not exist', () => {
    const payload = { userId: 'some-short-v4-' };
    const message = { type: MessageType.MoveSkipped, payload };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock).not.toBeCalled();
  });

  it('should not skip a move, because user state is not ActionNotTaken', () => {
    const payload = { userId: 'some-short-v4-uuid-0' };
    const message = { type: MessageType.MoveSkipped, payload };
    playerService.newMessageReceived(ws, message);
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock).not.toBeCalled();
  });

  it('should remove a player', () => {
    const payload = { userId: 'some-short-v4-uuid-0' };
    const message = { type: MessageType.RemovePlayer, payload };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock).toBeCalled();
  });

  it('should not remove a player, because user do not exist', () => {
    const payload = { userId: 'some-short-v4-' };
    const message = { type: MessageType.RemovePlayer, payload };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock).not.toBeCalled();
  });

  it('should set a turn if user`s move already exist', () => {
    const sendMock = jest.spyOn(ws, 'send');
    playerService.checkIfUserAlreadyExists(ws);
    expect(sendMock.mock.calls).toMatchSnapshot();
  });

  it('should move a chess figure', () => {
    const payload = { move: 'test' };
    const message = { type: MessageType.FigureMoved, payload };
    const sendMock = jest.spyOn(ws, 'send');
    playerService.newMessageReceived(ws, message);
    expect(sendMock.mock.calls).toMatchSnapshot();
    expect(gameService.figureMoved).toBeCalledWith(payload);
  });

  it('should clear the board', async () => {
    const sendMock = jest.spyOn(ws, 'send');
    playerService.clearBoard();
    expect(sendMock.mock.calls).toMatchSnapshot();
  });

  it('should disconnect a player', async () => {
    const sendMock = jest.spyOn(ws, 'send');
    playerService.playerDisconnected();
    expect(sendMock.mock.calls).toMatchSnapshot();
  });

  it('should unsubscribe twice without crash', async () => {
    playerService.unsubscribe(ws);
    playerService.unsubscribe(ws);
  });
});
