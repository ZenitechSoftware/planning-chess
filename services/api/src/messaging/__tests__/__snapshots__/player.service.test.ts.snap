// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player.service should clear the board 1`] = `
Array [
  Array [
    "{\\"type\\":\\"ClearBoard\\",\\"payload\\":[]}",
  ],
  Array [
    "{\\"type\\":\\"NewBoardState\\"}",
  ],
  Array [
    "{\\"type\\":\\"UpdatePlayerList\\",\\"payload\\":[{\\"status\\":\\"ActionNotTaken\\"}]}",
  ],
]
`;

exports[`player.service should disconnect a player 1`] = `
Array [
  Array [
    "{\\"type\\":\\"PlayerDisconnected\\",\\"payload\\":[{\\"status\\":\\"ActionNotTaken\\"}]}",
  ],
]
`;

exports[`player.service should join a new player 1`] = `
Array [
  Array [
    "{\\"type\\":\\"UpdatePlayerList\\",\\"payload\\":[{\\"id\\":\\"some-short-v4-uuid-0\\",\\"name\\":\\"foo\\",\\"color\\":{\\"background\\":{\\"r\\":256,\\"g\\":256,\\"b\\":256},\\"text\\":{\\"r\\":128,\\"g\\":128,\\"b\\":128}},\\"status\\":\\"ActionNotTaken\\"}]}",
  ],
  Array [
    "{\\"type\\":\\"NewBoardState\\"}",
  ],
]
`;

exports[`player.service should move a chess figure 1`] = `
Array [
  Array [
    "{\\"type\\":\\"FigureMoved\\"}",
  ],
  Array [
    "{\\"type\\":\\"UpdatePlayerList\\",\\"payload\\":[{\\"status\\":\\"FigurePlaced\\"}]}",
  ],
  Array [
    "{\\"type\\":\\"NewBoardState\\"}",
  ],
]
`;

exports[`player.service should set a turn if user\`s move already exist 1`] = `Array []`;

exports[`player.service should skip a move for a player 1`] = `
Array [
  Array [
    "{\\"type\\":\\"MoveSkipped\\",\\"payload\\":[{\\"id\\":\\"some-short-v4-uuid-0\\",\\"name\\":\\"foo\\",\\"color\\":{\\"background\\":{\\"r\\":256,\\"g\\":256,\\"b\\":256},\\"text\\":{\\"r\\":128,\\"g\\":128,\\"b\\":128}},\\"status\\":\\"MoveSkipped\\"}]}",
  ],
  Array [
    "{\\"type\\":\\"NewBoardState\\"}",
  ],
]
`;
